File: Angular5-Store/src/app/reducers/app/AppReducer.ts
import { ActionReducer, Action } from '@ngrx/store'
import UserState from './UserState'
import ToolbarState from './ToolbarState'
import SideNavState from './SideNavState'
import PageState from './PageState'
import DialogState from './DialogState'
import FabState from './FabState'

const reducerName = 'AppReducer@@: '
const SET_USER: string = reducerName + 'set user state'
const SET_TOOLBAR: string = reducerName + 'set toolbar state'
const SET_SIDENAV: string = reducerName + 'set side nav state'
const SET_SELECTED_PAGE: string = reducerName + 'set selected page state'
const SET_LOADING: string = reducerName + 'set loading state'
const TOGGLE_SIDE_NAV_MENU: string = reducerName + 'toggle side nav menu'
const SET_SIDE_NAV_MENU: string = reducerName + 'set side nav menu open or close'
const SET_PAGE: string = reducerName + 'set page state'
const SET_DIALOG: string = reducerName + 'set dialog state'

export class SET_USER_ACTION implements Action {
  readonly type: string = SET_USER
  constructor(public payload: UserState) { }
}
export class SET_TOOLBAR_ACTION implements Action {
  readonly type: string = SET_TOOLBAR
  constructor(public payload: ToolbarState) { }
}
export class SET_SIDENAV_ACTION implements Action {
  readonly type: string = SET_SIDENAV
  constructor(public payload: SideNavState) { }
}
export class SET_SIDE_NAV_MENU_ACTION implements Action {
  readonly type: string = SET_SIDE_NAV_MENU
  constructor(public payload: boolean) { }
}
export class SET_SELECTED_PAGE_ACTION implements Action {
  readonly type: string = SET_SELECTED_PAGE
  constructor(public payload: string) { }
}
export class SET_LOADING_ACTION implements Action {
  readonly type: string = SET_LOADING
  constructor(public payload: boolean) { }
}
export class SET_PAGE_ACTION implements Action {
  readonly type: string = SET_PAGE
  constructor(public payload: PageState) { }
}
export class TOGGLE_SIDE_NAV_MENU_ACTION implements Action {
  readonly type: string = TOGGLE_SIDE_NAV_MENU
}
export class SET_DIALOG_ACTION implements Action {
  readonly type: string = SET_DIALOG
  timeStamp: number = new Date().getDate() // give a new time stamp
  constructor(public payload: DialogState) { }
}

export interface AppState {
  dialog: DialogState
  sideNav: SideNavState
  fab: FabState,
  toolbar: ToolbarState
  user: UserState
  page: PageState
  selectedPage: string
  isloading: boolean
}

const defaultState: AppState = {
  dialog: new DialogState(),
  sideNav: new SideNavState(),
  fab: new FabState(),
  toolbar: new ToolbarState('Day'),
  user: new UserState(),
  page: new PageState(),
  selectedPage: '',
  isloading: false
}

export function AppReducer(state: AppState = defaultState, action: Action) {
  switch (action.type) {
    case SET_USER: return { ...state, user: (action as SET_USER_ACTION).payload }
    case SET_TOOLBAR: return { ...state, toolbar: (action as SET_TOOLBAR_ACTION).payload }
    case SET_SIDENAV: return { ...state, sideNav: (action as SET_SIDENAV_ACTION).payload }
    case SET_SELECTED_PAGE:
      var newState = { ...state }
      newState.sideNav.selectedid = (action as SET_SELECTED_PAGE_ACTION).payload
      newState.toolbar.title = state.sideNav.menuList.find(menuItem => menuItem.id == newState.sideNav.selectedid).title
      if (newState.page.isMobile) {
        newState.sideNav.isOpen = false
      }
      return newState
    case SET_LOADING: return { ...state, isloading: (action as SET_LOADING_ACTION).payload }
    case TOGGLE_SIDE_NAV_MENU: return { ...state, sideNav: { ...state.sideNav, isOpen: !state.sideNav.isOpen } }
    case SET_SIDE_NAV_MENU: return { ...state, sideNav: { ...state.sideNav, isOpen: (action as SET_SIDE_NAV_MENU_ACTION).payload } }
    case SET_PAGE:
      var page = (action as SET_PAGE_ACTION).payload
      var newState = { ...state }
      newState.page = page
      newState.sideNav.isOpen = newState.sideNav.isOpen ? page.width < 500 ? false : true : page.width > 1000 && page.width ? true : false
      return newState
    case SET_DIALOG: return { ...state, dialog: (action as SET_DIALOG_ACTION).payload }
    default:
      return state
  }
}

File: Android-Kotlin-Store/app/src/main/java/android/kotlinstore/Reducers/App/AppReducer.kt
package android.kotlinstore.Reducers.App

import android.kotlinstore.Store.Action
import android.kotlinstore.Store.Reducer

class AppReducer : Reducer() {
    companion object {
        val name = "AppReducer"
        val LOGIN_PAGE = 0
        val HOME_PAGE = 1
        val SETTING_PAGE = 2
        val ADD_TO_DAY_PAGE = 3

        private val reducerName = "[$name:] "

        private val SET_USER: String = reducerName + "set user user"
        private val SET_DRAW: String = reducerName + "set draw state"

        private val OPEN_FAB_MENU: String = reducerName + "open fab menu"
        private val CLOSE_FAB_MENU: String = reducerName + "close fab menu"
        private val TOGGLE_FAB_MENU: String = reducerName + "toggle fab menu"
        private val SET_FAB_ICON: String = reducerName + "set fab icon"
        private val SET_FAB_STATE: String = reducerName + "set fab state"

        private val SEND_SNACKBAR_MESSAGE: String = reducerName + "open up a snackbar message"

        private val SET_HOME_PAGE: String = reducerName + "set home page"
        private val SET_LOGIN_PAGE: String = reducerName + "set login page"
        private val SET_SETTINGS_PAGE: String = reducerName + "set settings page"
        private val SET_ADD_TO_DAY_PAGE: String = reducerName + "set add to day page"

        private val SET_LOADING: String = reducerName + "set loading state"

        private val SET_DIALOG: String = reducerName + "set dialog state"
    }

    data class State (
            var dialog: DialogState = DialogState(),
            var draw: DrawState = DrawState(),
            var fab: FabState = FabState(),
            var snack: SnackState = SnackState(),
            var toolbar: ToolbarState = ToolbarState(),
            var user: UserState = UserState(),
            var page: PageState = PageState(),
            var isLoading: Boolean = false
    )

    class SET_USER_ACTION(payload: UserState) : Action(type= SET_USER) { var payload = payload }
    class SET_DRAW_ACTION(payload: DrawState) : Action(type= SET_DRAW) { var payload = payload }

    class SET_FAB_ICON_ACTION(payload: Int) : Action(type=SET_FAB_ICON) { var payload = payload }
    class SET_FAB_STATE_ACTION(payload: FabState) : Action(type=SET_FAB_STATE) { var payload = payload }

    class SEND_SNACKBAR_MESSAGE_ACTION(payload: SnackState) : Action(type=SEND_SNACKBAR_MESSAGE) { var payload = payload }

    class SET_LOGIN_PAGE_ACTION() : Action(type=SET_LOGIN_PAGE)
    class SET_HOME_PAGE_ACTION() : Action(type=SET_HOME_PAGE)
    class SET_SETTINGS_PAGE_ACTION() : Action(type=SET_SETTINGS_PAGE)
    class SET_ADD_TO_DAY_PAGE_ACTION() : Action(type=SET_ADD_TO_DAY_PAGE)

    class SET_LOADING_ACTION(payload: Boolean) : Action(type=SET_LOADING) { var payload = payload }
    class SET_DIALOG_ACTION(payload: DialogState) : Action(type= SET_DIALOG) { var payload = payload }

    override fun reduce(storeState: Any?, action: Action):Any? {
        var state: State? = storeState as State?
        if (null == state) {
            state = State()
        }

        return when(action.type) {
            SET_USER ->state.copy(user = (action as SET_USER_ACTION).payload)
            SET_DRAW ->state.copy(draw = (action as SET_DRAW_ACTION).payload)

            OPEN_FAB_MENU->state.copy(fab = state.fab.copy(isOpen = true))
            CLOSE_FAB_MENU->state.copy(fab = state.fab.copy(isOpen = false))
            TOGGLE_FAB_MENU->state.copy(fab = state.fab.copy(isOpen = !state.fab.isOpen))

            SET_FAB_ICON ->state.copy(fab = state.fab.copy(icon = (action as SET_FAB_ICON_ACTION).payload))
            SET_FAB_STATE ->state.copy(fab = (action as SET_FAB_STATE_ACTION).payload)

            SEND_SNACKBAR_MESSAGE ->state.copy(snack = (action as SEND_SNACKBAR_MESSAGE_ACTION).payload)

            SET_LOGIN_PAGE ->state.copy(
                    toolbar = ToolbarState(
                        title = "Login page",
                        showToolbar = false,
                        showBackBtn = false
                    ),
                    fab = FabState(showFab = false, isOpen = false),
                    page = PageState(selectedPage = LOGIN_PAGE)
            )
            SET_HOME_PAGE ->state.copy(
                    toolbar = ToolbarState(
                        title = "Home page",
                        showToolbar = true,
                        showBackBtn = false
                    ),
                    fab = FabState(showFab = true, isOpen = false),
                    page = PageState(selectedPage = HOME_PAGE)
            )
            SET_SETTINGS_PAGE->state.copy(
                    toolbar = ToolbarState(
                            title = "Settings page",
                            showToolbar = true,
                            showBackBtn = false
                    ),
                    fab = FabState(showFab = false, isOpen = false),
                    page = PageState(selectedPage = SETTING_PAGE)
            )
            SET_ADD_TO_DAY_PAGE->state.copy(
                    toolbar = ToolbarState(
                            title = "Add to Day page",
                            showToolbar = true,
                            showBackBtn = true
                    ),
                    fab = FabState(showFab = true, isOpen = false),
                    page = PageState(selectedPage = ADD_TO_DAY_PAGE)
            )

            SET_LOADING->state.copy(isLoading = (action as SET_LOADING_ACTION).payload)
            SET_DIALOG ->state.copy(dialog = (action as SET_DIALOG_ACTION).payload)
            else->state
        }
    }
}